File: packages/swift-client/Sources/PyricFirestore/Transport/BridgeOperations.swift
import Foundation

extension PyricBridgeClient {

    /// Reads a document snapshot via getDoc.
    public func getDoc(path: String, actAs: AuthLens? = nil) async throws -> AnySendable {
        try await op(
            method: "getDoc",
            params: ["path": .string(path)],
            actAs: actAs
        )
    }

    /// Reads query documents via getDocs.
    public func getDocs(source: TargetDescriptor, actAs: AuthLens? = nil) async throws -> AnySendable {
        try await op(
            method: "getDocs",
            params: ["source": source.toAnySendable()],
            actAs: actAs
        )
    }

    /// Writes document data via setDoc.
    public func setDoc(
        path: String,
        data: AnySendable,
        options: SetOptionsWire? = nil,
        actAs: AuthLens? = nil
    ) async throws {
        var params: [String: AnySendable] = [
            "path": .string(path),
            "data": data
        ]
        if let options {
            params["options"] = options.toAnySendable()
        }
        _ = try await op(method: "setDoc", params: params, actAs: actAs)
    }

    /// Writes document data via setDoc with a Swift dictionary payload.
    public func setDoc(
        path: String,
        data: [String: Any],
        options: SetOptionsWire? = nil,
        actAs: AuthLens? = nil
    ) async throws {
        let encoded = try ValueCodec.encodeWriteData(data)
        try await setDoc(path: path, data: AnySendable.from(encoded), options: options, actAs: actAs)
    }

    /// Updates document fields via updateDoc.
    public func updateDoc(
        path: String,
        data: AnySendable,
        actAs: AuthLens? = nil
    ) async throws {
        _ = try await op(
            method: "updateDoc",
            params: [
                "path": .string(path),
                "data": data
            ],
            actAs: actAs
        )
    }

    /// Updates document fields via updateDoc with a Swift dictionary payload.
    public func updateDoc(
        path: String,
        data: [String: Any],
        actAs: AuthLens? = nil
    ) async throws {
        let encoded = try ValueCodec.encodeWriteData(data)
        try await updateDoc(path: path, data: AnySendable.from(encoded), actAs: actAs)
    }

    /// Deletes a document via deleteDoc.
    public func deleteDoc(path: String, actAs: AuthLens? = nil) async throws {
        _ = try await op(
            method: "deleteDoc",
            params: ["path": .string(path)],
            actAs: actAs
        )
    }

    /// Creates a document with an auto-minted ID under collectionPath via addDoc.
    public func addDoc(
        collectionPath: String,
        data: AnySendable,
        actAs: AuthLens? = nil
    ) async throws -> (id: String, path: String) {
        let res = try await op(
            method: "addDoc",
            params: [
                "collectionPath": .string(collectionPath),
                "data": data
            ],
            actAs: actAs
        )
        guard let id = res["id"]?.stringValue,
              let path = res["path"]?.stringValue else {
            throw PyricBridgeError.internalError("addDoc response missing id or path: \(res)")
        }
        return (id: id, path: path)
    }

    /// Creates a document with an auto-minted ID under collectionPath via addDoc with a Swift dictionary payload.
    public func addDoc(
        collectionPath: String,
        data: [String: Any],
        actAs: AuthLens? = nil
    ) async throws -> (id: String, path: String) {
        let encoded = try ValueCodec.encodeWriteData(data)
        return try await addDoc(collectionPath: collectionPath, data: AnySendable.from(encoded), actAs: actAs)
    }

    /// Counts matching documents via count.
    public func count(source: TargetDescriptor, actAs: AuthLens? = nil) async throws -> Int {
        let res = try await op(
            method: "count",
            params: ["source": source.toAnySendable()],
            actAs: actAs
        )
        return Int(res["count"]?.intValue ?? 0)
    }

    /// Performs server-side aggregations via aggregate.
    public func aggregate(
        source: TargetDescriptor,
        spec: [String: AggregateFieldDescriptor],
        actAs: AuthLens? = nil
    ) async throws -> [String: Double?] {
        let specMap = spec.mapValues { $0.toAnySendable() }
        let res = try await op(
            method: "aggregate",
            params: [
                "source": source.toAnySendable(),
                "spec": .dictionary(specMap)
            ],
            actAs: actAs
        )
        guard let data = res["data"]?.dictionaryValue else {
            return [:]
        }
        var result: [String: Double?] = [:]
        for (k, v) in data {
            if v == .null {
                result[k] = nil
            } else {
                result[k] = v.doubleValue
            }
        }
        return result
    }

    /// Atomically commits a batch of write mutations via batchCommit.
    public func batchCommit(writes: [WriteDescriptor], actAs: AuthLens? = nil) async throws {
        _ = try await op(
            method: "batchCommit",
            params: ["writes": .array(writes.map { $0.toAnySendable() })],
            actAs: actAs
        )
    }

    /// Commits an interactive transaction via txnCommit.
    public func txnCommit(
        reads: [TxnReadEntry],
        writes: [WriteDescriptor],
        actAs: AuthLens? = nil
    ) async throws {
        _ = try await op(
            method: "txnCommit",
            params: [
                "reads": .array(reads.map { $0.toAnySendable() }),
                "writes": .array(writes.map { $0.toAnySendable() })
            ],
            actAs: actAs
        )
    }
}

File: packages/flutter-client/lib/src/transport/bridge_operations.dart
import 'bridge_client.dart';
import 'codecs.dart';

/// Extension methods adding Firestore-specific RPC conveniences to [PyricBridgeClient].
extension PyricBridgeFirestoreOps on PyricBridgeClient {
  /// Reads a document snapshot via `getDoc`.
  Future<dynamic> getDoc(String path, {Map<String, dynamic>? actAs}) {
    return op('getDoc', {'path': path}, actAs: actAs);
  }

  /// Reads query documents via `getDocs`.
  Future<dynamic> getDocs(
    Map<String, dynamic> source, {
    Map<String, dynamic>? actAs,
  }) {
    return op('getDocs', {'source': source}, actAs: actAs);
  }

  /// Writes document data via `setDoc`.
  Future<dynamic> setDoc(
    String path,
    Map<String, dynamic> data, {
    Map<String, dynamic>? options,
    Map<String, dynamic>? actAs,
  }) {
    final params = <String, dynamic>{
      'path': path,
      'data': encodeWriteData(data),
      if (options != null) 'options': options,
    };
    return op('setDoc', params, actAs: actAs);
  }

  /// Updates document fields via `updateDoc`.
  Future<dynamic> updateDoc(
    String path,
    Map<String, dynamic> data, {
    Map<String, dynamic>? actAs,
  }) {
    return op(
      'updateDoc',
      {
        'path': path,
        'data': encodeWriteData(data),
      },
      actAs: actAs,
    );
  }

  /// Deletes a document via `deleteDoc`.
  Future<dynamic> deleteDoc(String path, {Map<String, dynamic>? actAs}) {
    return op('deleteDoc', {'path': path}, actAs: actAs);
  }

  /// Creates a document with auto-ID under collection via `addDoc`.
  Future<dynamic> addDoc(
    String collectionPath,
    Map<String, dynamic> data, {
    Map<String, dynamic>? actAs,
  }) {
    return op(
      'addDoc',
      {
        'collectionPath': collectionPath,
        'data': encodeWriteData(data),
      },
      actAs: actAs,
    );
  }

  /// Counts matching documents via `count`.
  Future<int> count(
    Map<String, dynamic> source, {
    Map<String, dynamic>? actAs,
  }) async {
    final res = await op('count', {'source': source}, actAs: actAs);
    if (res is Map && res.containsKey('count')) {
      return (res['count'] as num).toInt();
    }
    return 0;
  }

  /// Performs server-side aggregations via `aggregate`.
  Future<Map<String, dynamic>> aggregate(
    Map<String, dynamic> source,
    Map<String, dynamic> spec, {
    Map<String, dynamic>? actAs,
  }) async {
    final res = await op(
      'aggregate',
      {
        'source': source,
        'spec': spec,
      },
      actAs: actAs,
    );
    if (res is Map && res.containsKey('data') && res['data'] is Map) {
      return Map<String, dynamic>.from(res['data'] as Map);
    }
    return <String, dynamic>{};
  }

  /// Atomically commits a batch of write mutations via `batchCommit`.
  Future<dynamic> batchCommit(
    List<Map<String, dynamic>> writes, {
    Map<String, dynamic>? actAs,
  }) {
    return op('batchCommit', {'writes': writes}, actAs: actAs);
  }

  /// Commits an interactive transaction via `txnCommit`.
  Future<dynamic> txnCommit(
    List<Map<String, dynamic>> reads,
    List<Map<String, dynamic>> writes, {
    Map<String, dynamic>? actAs,
  }) {
    return op(
      'txnCommit',
      {
        'reads': reads,
        'writes': writes,
      },
      actAs: actAs,
    );
  }
}
